### PR TITLE
Support non-ASCII encodings in Content-Disposition field per RFC 5987 and RFC 6266

### PR DESCRIPTION
**Pull request by stanhu**
See: http://greenbytes.de/tech/tc2231/

### CODE REVIEW

File: actionpack/lib/action_controller/metal/data_streaming.rb [hunk 1]
```diff
@@ -155,7 +156,17 @@ def send_file_headers!(options)
...
-          disposition += %(; filename="#{options[:filename]}") if options[:filename]
+          filename = options[:filename]
+          if filename
+            disposition += %(; filename="#{filename}")
+
+            # Support RFC 6266
+            # http://tools.ietf.org/html/rfc6266#appendix-D
+            unless filename.ascii_only?
+              encoded_filename = CGI::escape(filename)
+              disposition += %(; filename*="#{encoded_filename.encoding}''#{encoded_filename}")
```
Comment (matthewd) [inline actionpack/lib/action_controller/metal/data_streaming.rb]: I don't think these quotes should be here. Also, does `CGI.escape` cover all the characters we need it to?

**Comment by matthewd** [comment]
We'll need some tests, too.

File: actionpack/lib/action_controller/metal/data_streaming.rb (hunk 1 above)
Comment (icambron) [inline actionpack/lib/action_controller/metal/data_streaming.rb]: Yeah, no quotes. `CGI::escape` should do all the characters, but also a bunch that it shouldn't. That might not be a problem in practice, but this part is a definitely problem:

```
irb(main):004:0> CGI::escape("foo bar")
=> "foo+bar"
```

where it should use `%20`. I have a library that does this stuff in Clojure, and it a) checks that the character is specifically excluded by the spec before passing it to the escaper and b) "manually" replaces the space with the right escape sequence. You can see the relevant code here if you grok CLJ: https://github.com/Zensight/header-utils/blob/master/src/header_utils/encoding.clj#L43

Comment (jeremy) [inline actionpack/lib/action_controller/metal/data_streaming.rb]: Here's what we're using:

```ruby
headers['Content-Disposition'] = "#{disposition}; #{content_disposition_filename(filename)}"

# RFC2231 encoding for UTF-8 filenames, with an ASCII fallback
# first for unsupported browsers (IE < 9, perhaps others?).
# http://greenbytes.de/tech/tc2231/#encoding-2231-fb
def content_disposition_filename(filename)
  ascii_filename = encode_ascii_filename(filename)
  utf8_filename = encode_utf8_filename(filename)
  "#{ascii_filename}; #{utf8_filename}"
end

TRADITIONAL_PARAMETER_ESCAPED_CHAR = /[^ A-Za-z0-9!#$+.^_`|~-]/

def encode_ascii_filename(filename)
  # There is no reliable way to escape special or non-Latin characters
  # in a traditionally quoted Content-Disposition filename parameter.
  # Settle for transliterating to ASCII, then percent-escaping special
  # characters, excluding spaces.
  filename = I18n.transliterate(filename)
  filename = percent_escape(filename, TRADITIONAL_PARAMETER_ESCAPED_CHAR)
  %(filename="#{filename}")
end

RFC5987_PARAMETER_ESCAPED_CHAR = /[^A-Za-z0-9!#$&+.^_`|~-]/

def encode_utf8_filename(filename)
  # RFC2231 filename parameters can simply be percent-escaped according
  # to RFC5987.
  filename = percent_escape(filename, RFC5987_PARAMETER_ESCAPED_CHAR)
  %(filename*=UTF-8''#{filename})
end

def percent_escape(string, pattern)
  string.gsub(pattern) do |char|
    char.bytes.map { |byte| "%%%02X" % byte }.join("")
  end
end
```

**Comment by icambron** [comment]
IMO, it's a bit odd to support 6266 without also supporting quoted strings for ascii-but-reserved characters (also part of 5987). It creates a weird, non-working middle ground between vanilla file names that always worked and non-ascii chars that will now work. "Your file download isn't parsing correctly? Just add a non-ascii char to the name and it work!"

File: actionpack/lib/action_controller/metal/data_streaming.rb (hunk 1 above)
Comment (icambron) [inline actionpack/lib/action_controller/metal/data_streaming.rb]: @jeremy Why are you not escaping spaces? The spec says:

```
value-chars   = *( pct-encoded / attr-char )

...

attr-char     = ALPHA / DIGIT
                   / "!" / "#" / "$" / "&" / "+" / "-" / "."
                   / "^" / "_" / "`" / "|" / "~"
                  ; token except ( "*" / "'" / "%" )
```

Which implies to me that spaces are excluded - they're not in that list and and 2616 specifically excludes them from the definition of "token":

```
      token          = 1*<any CHAR except CTLs or separators>
      separators     = "(" | ")" | "<" | ">" | "@"
                      | "," | ";" | ":" | "\" | <">
                      | "/" | "[" | "]" | "?" | "="
                      | "{" | "}" | SP | HT
```

But if I'm missing something, I'd like to change my implementation too!

Comment (jeremy) [inline actionpack/lib/action_controller/metal/data_streaming.rb]: @icambron Note we use 5987 for `filename*` only and do %-escape there. We use an old-style `quoted-string` for the ASCII `filename`, where spaces are allowed, for wide compat with older browsers.

Comment (icambron) [inline actionpack/lib/action_controller/metal/data_streaming.rb]: @jeremy - I deleted my comment after figuring that out :)

**Comment by dup2** [comment]
+1

**Comment by ypresto** [comment]
We are also facing this issue. We're going to add below snippet to ApplicationController:
http://qiita.com/rorensu2236/items/638d181e155b5dc5e35c#comment-c64bd5cf57177830bdcc

```rb
  # NOTE: This monkey patch is for UTF-8 file name in Content-Disposition header.
  # Based on http://qiita.com/rorensu2236/items/638d181e155b5dc5e35c
  def send_file_headers!(options)
    super(options)
    match = /(.+); filename="(.+)"/.match(headers['Content-Disposition'])
    return unless match
    # macOS uses nfd format. Normalize it with nfc for Windows compatibility.
    encoded = URI. encode_www_form_component(ActiveSupport::Multibyte::Unicode.normalize(match[2], :c))
    # TODO: PR rails; transliterate's replacement parameter does not applied for Japanese letters.
    ascii = I18n.transliterate(match[2], '-').tr('?', '-').gsub(/[\\"]/, { '"' => '\\"', '\\' => '\\\\' })
    headers['Content-Disposition'] = %(#{match[1]}; filename="#{ascii}"; filename*=UTF-8''#{encoded}) unless encoded == match[2]
  end
```

----

I think it'd be better to have default logic for UTF-8 filename, while it might not (maybe can't) be perfect.

**Comment by ypresto** [comment]
We are also facing this issue. We're going to add below snippet to ApplicationController:
http://qiita.com/rorensu2236/items/638d181e155b5dc5e35c#comment-c64bd5cf57177830bdcc

```rb
  # NOTE: This monkey patch is for UTF-8 file name in Content-Disposition header.
  # Based on http://qiita.com/rorensu2236/items/638d181e155b5dc5e35c
  def send_file_headers!(options)
    super(options)
    match = /(.+); filename="(.+)"/.match(headers['Content-Disposition'])
    return unless match
    # macOS uses nfd format. Normalize it with nfc for Windows compatibility.
    encoded = ERB::Util.url_encode(ActiveSupport::Multibyte::Unicode.normalize(match[2], :c))
    ascii = ActiveSupport::Inflector.transliterate(match[2], '-').gsub(/[\\"]/, { '"' => '\\"', '\\' => '\\\\' })
    headers['Content-Disposition'] = %(#{match[1]}; filename="#{ascii}"; filename*=UTF-8''#{encoded}) unless encoded == match[2]
  end
```

Note that ERB::Util.url_encode encodes half-width white space into `%20`.
http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/29373

----

I think it'd be better to have default logic for UTF-8 filename, while it might not (maybe can't) be perfect.

File: actionpack/lib/action_controller/metal/data_streaming.rb (hunk 1 above)
Comment (mtsmfm) [reply actionpack/lib/action_controller/metal/data_streaming.rb]: @jeremy Do you have any plan to apply your patch to Rails?

Comment (jeremy) [reply actionpack/lib/action_controller/metal/data_streaming.rb]: Please do, @mtsmfm. Note that Active Storage adopted this as well (https://github.com/rails/rails/blob/master/activestorage/app/models/active_storage/filename/parameters.rb) so perhaps its implementation could be shared and reused.

Comment (mtsmfm) [reply actionpack/lib/action_controller/metal/data_streaming.rb]: @jeremy sent a PR https://github.com/rails/rails/pull/33829

**Comment by mtsmfm** [comment]
This PR can be closed because https://github.com/rails/rails/pull/33829 got merged

**Pull request closed without merging**